### PR TITLE
Handle cases tables preceded by complex queries

### PR DESCRIPTION
**Pull request by Byunk**
#422

### CODE REVIEW

**Comment by macbre** [comment]
Thanks for you PR, @Byunk! And adding test cases.

**Review by macbre — APPROVED** [review]

File: sql_metadata/parser.py [hunk 1]
```diff
@@ -970,6 +970,9 @@ def replace_back_quotes_in_string(match):
         return query
 
     def _determine_last_relevant_keyword(self, token: SQLToken, last_keyword: str):
+        if token.value == "," and token.last_keyword_normalized == "ON":
+            return "FROM"
+
```
Comment (macbre) [inline sql_metadata/parser.py:975]:
```
poetry run flake8 sql_metadata
sql_metadata/parser.py:975:1: W293 blank line contains whitespace
```

Comment (macbre) [reply sql_metadata/parser.py:975]:
```suggestion

```

File: test/test_getting_tables.py [hunk 2]
```diff
@@ -666,3 +666,76 @@ def test_insert_into_on_duplicate_key_ipdate():
         " VALUES ('user1', 'john doe', 20)"
         " ON DUPLICATE KEY UPDATE name='john doe', age=20"
     ).tables == ["user"]
+
+
```
Comment (macbre) [inline test/test_getting_tables.py:670]: @Byunk - can you please reformat that file using `black`?

File: test/test_getting_tables.py [hunk 3]
```diff
@@ -666,3 +666,76 @@ def test_insert_into_on_duplicate_key_ipdate():
...
+    parser = Parser("""
+    SELECT * 
+    FROM Sales 
+        JOIN Customers 
+            ON Sales.CustomerID = Customers.CustomerID, 
+        (SELECT MAX(Revenue) FROM Sales),
+        Stores
+    """)
+    assert parser.tables == ["Sales", "Customers", "Stores"]
```
Comment (macbre) [inline test/test_getting_tables.py:741]:
```suggestion
def test_join_followed_by_tables():
    query = """
    SELECT 
        web_site_id, 
        sum(sales_price) AS sales, 
        sum(profit) AS profit, 
        sum(return_amt) AS returns1, 
        sum(net_loss) AS profit_loss 
    FROM 
        (
        SELECT 
            ws_web_site_sk AS wsr_web_site_sk, 
            ws_sold_date_sk AS date_sk, 
            ws_ext_sales_price AS sales_price, 
            ws_net_profit AS profit, 
            cast(
            0 AS decimal(7, 2)
            ) AS return_amt, 
            cast(
            0 AS decimal(7, 2)
            ) AS net_loss 
        FROM 
            web_sales 
        UNION ALL 
        SELECT 
            ws_web_site_sk AS wsr_web_site_sk, 
            wr_returned_date_sk AS date_sk, 
            cast(
            0 AS decimal(7, 2)
            ) AS sales_price, 
            cast(
            0 AS decimal(7, 2)
            ) AS profit, 
            wr_return_amt AS return_amt, 
            wr_net_loss AS net_loss 
        FROM 
            web_returns 
            LEFT OUTER JOIN web_sales ON (
            wr_item_sk = ws_item_sk 
            AND wr_order_number = ws_order_number
            )
        ) salesreturns, 
        date_dim, 
        web_site 
    WHERE 
        date_sk = d_date_sk 
        AND d_date BETWEEN cast('2002-08-22' AS date) 
        AND (
        cast('2002-08-22' AS date) + INTERVAL '14' day
        ) 
        AND wsr_web_site_sk = web_site_sk 
    GROUP BY 
        web_site_id
    """
    parser = Parser(query)
    assert parser.tables == ["web_sales", "web_returns", "date_dim", "web_site"]

    parser = Parser(
        """
    SELECT * 
    FROM Sales 
        JOIN Customers 
            ON Sales.CustomerID = Customers.CustomerID, 
        (SELECT MAX(Revenue) FROM Sales),
        Stores
    """
    )
    assert parser.tables == ["Sales", "Customers", "Stores"]
```

File: test/test_getting_tables.py (hunk 2 above)
Comment (Byunk) [reply test/test_getting_tables.py:670]: @macbre sure. please check it!

**Comment by Byunk** [comment]
Hi @macbre. I;ve got another bug. The `customer` table, which is followed by subquery is not detected by parser. I think this is because the parser does not store keyword that is preceded by subquery in the FROM statement. So I added another testcase and list of token, `_preceded_keywords` to solve this issue. It passed all the testcases but I'm not that sure there won't be any side effect though. Could you please check?

## Tested query

```sql
SELECT top 100 c_last_name ,
           c_first_name ,
           ca_city ,
           bought_city ,
           ss_ticket_number ,
           amt,
           profit
    FROM
    (SELECT ss_ticket_number ,
            ss_customer_sk ,
            ca_city bought_city ,
            sum(ss_coupon_amt) amt ,
            sum(ss_net_profit) profit
    FROM store_sales,
            date_dim,
            store,
            household_demographics,
            customer_address
    WHERE store_sales.ss_sold_date_sk = date_dim.d_date_sk
        AND store_sales.ss_store_sk = store.s_store_sk
        AND store_sales.ss_hdemo_sk = household_demographics.hd_demo_sk
        AND store_sales.ss_addr_sk = customer_address.ca_address_sk
        AND (household_demographics.hd_dep_count = 3
            OR household_demographics.hd_vehicle_count= 2)
        AND date_dim.d_dow IN (6,
                                0)
        AND date_dim.d_year IN (1998,
                                1998+1,
                                1998+2)
        AND store.s_city IN ('Oak Grove',
                            'Fairview',
                            'Riverside',
                            'Five Points',
                            'Midway')
    GROUP BY ss_ticket_number,
                ss_customer_sk,
                ss_addr_sk,
                ca_city) dn,
        customer,
        customer_address current_addr
    WHERE ss_customer_sk = c_customer_sk
    AND customer.c_current_addr_sk = current_addr.ca_address_sk
    AND current_addr.ca_city <> bought_city
    ORDER BY c_last_name ,
            c_first_name ,
            ca_city ,
            bought_city ,
            ss_ticket_number
```

**Review by macbre — APPROVED** [review]
Thanks!

File: sql_metadata/parser.py [hunk 4]
```diff
@@ -164,6 +165,8 @@ def tokens(self) -> List[SQLToken]:
             elif token.is_right_parenthesis:
                 token.token_type = TokenType.PARENTHESIS
                 self._determine_closing_parenthesis_type(token=token)
+                if token.is_subquery_end:
+                    last_keyword = self._preceded_keywords.pop()
```
Comment (macbre) [inline sql_metadata/parser.py:170]: Now the linter complains about the complexity of that function.

**Comment by macbre** [comment]
@Byunk - the PR is ready to be merged, thanks for your change!

**Pull request merged**